### ISSUE TITLE
Malformed message; no clientId in message or inferred from connection

### ISSUE THREAD

**Issue by stannedelchev** [ably/ably-ios#451]
After updating to the v0.8.2 via CocoaPods, I'm getting a `Malformed message; no clientId in message or inferred from connection` error in the callback of [presence.enter(nil)](https://github.com/ably/demo-mobile-ios/blob/6627fd95670b5f393041b04dd9123cebece3243c/demo-ios/ChatModel.swift#L88).

Here's the log:
```
2016-05-20 09:28:14.499 demo-ios[7894:67011] DEBUG: (ARTRest.m:57) RS:0x7fd528d440c0 0x7fd528d81730 alloc HTTP
2016-05-20 09:28:14.499 demo-ios[7894:67011] DEBUG: (ARTAuth.m:45) RS:0x7fd528d440c0 validating ARTClientOptions: key=(null) token=(null) authUrl=https://www.ably.io/ably-auth/token-details/demos authMethod=GET hasAuthCallback=0
2016-05-20 09:28:14.499 demo-ios[7894:67011] DEBUG: (ARTAuth.m:66) RS:0x7fd528d440c0 setting up auth method Token with authUrl
2016-05-20 09:28:14.499 demo-ios[7894:67011] DEBUG: (ARTRest.m:73) RS:0x7fd528d440c0 initialized
2016-05-20 09:28:14.500 demo-ios[7894:67011] DEBUG: (ARTRealtime.m:82) R:0x7fd528da7d70 initialized with RS:0x7fd528d440c0
2016-05-20 09:28:14.500 demo-ios[7894:67011] DEBUG: (ARTRealtime.m:212) R:0x7fd528da7d70 transition to Connecting requested
2016-05-20 09:28:14.500 demo-ios[7894:67011] DEBUG: (ARTWebSocketTransport.m:56) R:0x0 WS:0x7fd528c9e8b0 alloc
2016-05-20 09:28:14.500 demo-ios[7894:67011] DEBUG: (ARTWebSocketTransport.m:91) R:0x7fd528da7d70 WS:0x7fd528c9e8b0 websocket connect
2016-05-20 09:28:14.500 demo-ios[7894:67011] DEBUG: (ARTWebSocketTransport.m:105) R:0x7fd528da7d70 WS:0x7fd528c9e8b0 connecting with token auth; authorising
2016-05-20 09:28:14.501 demo-ios[7894:67011] VERBOSE: RS:0x7fd528d440c0 ARTAuth: requesting new token.
2016-05-20 09:28:14.501 demo-ios[7894:67011] DEBUG: (ARTAuth.m:175) RS:0x7fd528d440c0 using authUrl (GET https://www.ably.io/ably-auth/token-details/demos?clientId=abcdefgh&ttl=3600.000000&capability=%7B%22*%22:%5B%22*%22%5D%7D&timestamp=1463725694.501228&format=msgpack)
2016-05-20 09:28:14.502 demo-ios[7894:67011] DEBUG: (ARTRest.m:138) RS:0x7fd528d440c0 executing request <NSMutableURLRequest: 0x7fd528c8f9b0> { URL: https://www.ably.io/ably-auth/token-details/demos?clientId=abcdefgh&ttl=3600.000000&capability=%7B%22*%22:%5B%22*%22%5D%7D&timestamp=1463725694.501228&format=msgpack }
2016-05-20 09:28:14.502 demo-ios[7894:67011] DEBUG: GET https://www.ably.io/ably-auth/token-details/demos?clientId=abcdefgh&ttl=3600.000000&capability=%7B%22*%22:%5B%22*%22%5D%7D&timestamp=1463725694.501228&format=msgpack
2016-05-20 09:28:14.502 demo-ios[7894:67011] VERBOSE: Headers {
    Accept = "application/json,application/x-msgpack";
    "X-Ably-Version" = "0.8";
}
2016-05-20 09:28:14.503 demo-ios[7894:67011] DEBUG: (ARTRestChannel.m:35) RS:0x7fd528f5c000 instantiating under 'mobile:chat'
2016-05-20 09:28:14.504 demo-ios[7894:67011] DEBUG: (ARTRealtimeChannel.m:533) R:0x7fd528da7d70 C:0x7fd528d706d0 already attaching
2016-05-20 09:28:14.505 demo-ios[7894:67011] DEBUG: (ARTRealtimeChannel.m:533) R:0x7fd528da7d70 C:0x7fd528d706d0 already attaching
2016-05-20 09:28:15.906 demo-ios[7894:67011] DEBUG: GET https://www.ably.io/ably-auth/token-details/demos?clientId=abcdefgh&ttl=3600.000000&capability=%7B%22*%22:%5B%22*%22%5D%7D&timestamp=1463725694.501228&format=msgpack: statusCode 200
2016-05-20 09:28:15.906 demo-ios[7894:67011] VERBOSE: Headers {
    "Access-Control-Allow-Origin" = "http://docs.ably.io";
    "Cache-Control" = "max-age=0, private, must-revalidate";
    Connection = close;
    "Content-Encoding" = gzip;
    "Content-Type" = "application/json; charset=utf-8";
    Date = "Fri, 20 May 2016 06:28:15 GMT";
    Etag = "W/\"c4f6c4395131b4d5c4c5008874e3f9c7\"";
    Server = Cowboy;
    Status = "200 OK";
    Vary = "Accept-Encoding";
    Via = "1.1 vegur";
    "X-Content-Type-Options" = nosniff;
    "X-Frame-Options" = SAMEORIGIN;
    "X-Request-Id" = "756ef4fa-3a78-49ad-a28a-f1ba9e0a15aa";
    "X-Runtime" = "0.766614";
    "X-Xss-Protection" = "1; mode=block";
}
2016-05-20 09:28:15.907 demo-ios[7894:67011] DEBUG: RS:0x7fd528d440c0 ARTAuth: authUrl response <NSHTTPURLResponse: 0x7fd528eaf7e0> { URL: https://www.ably.io/ably-auth/token-details/demos?clientId=abcdefgh&ttl=3600.000000&capability=%7B%22*%22:%5B%22*%22%5D%7D&timestamp=1463725694.501228&format=msgpack } { status code: 200, headers {
    "Access-Control-Allow-Origin" = "http://docs.ably.io";
    "Cache-Control" = "max-age=0, private, must-revalidate";
    Connection = close;
    "Content-Encoding" = gzip;
    "Content-Type" = "application/json; charset=utf-8";
    Date = "Fri, 20 May 2016 06:28:15 GMT";
    Etag = "W/\"c4f6c4395131b4d5c4c5008874e3f9c7\"";
    Server = Cowboy;
    Status = "200 OK";
    Vary = "Accept-Encoding";
    Via = "1.1 vegur";
    "X-Content-Type-Options" = nosniff;
    "X-Frame-Options" = SAMEORIGIN;
    "X-Request-Id" = "756ef4fa-3a78-49ad-a28a-f1ba9e0a15aa";
    "X-Runtime" = "0.766614";
    "X-Xss-Protection" = "1; mode=block";
} }
2016-05-20 09:28:15.907 demo-ios[7894:67011] VERBOSE: RS:0x7fd528d440c0 ARTJsonLikeEncoder<json> decoding '<7b0a2020 22746f6b 656e223a 20224932 455f4a51 2e41725f 614f5967 78747067 52615731 62726953 62355848 4863567a 5f386d6f 54446b45 6f5a3439 494e3336 30355663 6d445522 2c0a2020 226b6579 4e616d65 223a2022 4932455f 4a512e72 5f614f59 67222c0a 20202269 73737565 64223a20 31343633 37323536 39353636 332c0a20 20226578 70697265 73223a20 31343633 37363838 39353636 332c0a20 20226361 70616269 6c697479 223a2022 7b5c222a 5c223a5b 5c222a5c 225d7d22 2c0a2020 22636c69 656e7449 64223a20 222a220a 7d>'; got: {
    capability = "{\"*\":[\"*\"]}";
    clientId = "*";
    expires = 1463768895663;
    issued = 1463725695663;
    keyName = "I2E_JQ.r_aOYg";
    token = "I2E_JQ.Ar_aOYgxtpgRaW1briSb5XHHcVz_8moTDkEoZ49IN3605VcmDU";
}
2016-05-20 09:28:15.907 demo-ios[7894:67011] VERBOSE: RS:0x7fd528d440c0 ARTJsonLikeEncoder<json>: tokenFromDictionary {
    capability = "{\"*\":[\"*\"]}";
    clientId = "*";
    expires = 1463768895663;
    issued = 1463725695663;
    keyName = "I2E_JQ.r_aOYg";
    token = "I2E_JQ.Ar_aOYgxtpgRaW1briSb5XHHcVz_8moTDkEoZ49IN3605VcmDU";
}
2016-05-20 09:28:15.908 demo-ios[7894:67011] VERBOSE: RS:0x7fd528d440c0 ARTAuth: token request succeeded: ARTTokenDetails: token=I2E_JQ.Ar_aOYgxtpgRaW1briSb5XHHcVz_8moTDkEoZ49IN3605VcmDU clientId=* issued=2016-05-20 06:28:15 +0000 expires=2016-05-20 18:28:15 +0000
2016-05-20 09:28:15.909 demo-ios[7894:67011] DEBUG: (ARTWebSocketTransport.m:109) R:0x7fd528da7d70 WS:0x7fd528c9e8b0 authorised: ARTTokenDetails: token=I2E_JQ.Ar_aOYgxtpgRaW1briSb5XHHcVz_8moTDkEoZ49IN3605VcmDU clientId=* issued=2016-05-20 06:28:15 +0000 expires=2016-05-20 18:28:15 +0000 error: (null)
2016-05-20 09:28:15.909 demo-ios[7894:67011] DEBUG: (ARTWebSocketTransport.m:181) R:0x7fd528da7d70 WS:0x7fd528c9e8b0 url /?accessToken=I2E_JQ.Ar_aOYgxtpgRaW1briSb5XHHcVz_8moTDkEoZ49IN3605VcmDU&client_id=abcdefgh&echo=true&format=msgpack&v=0.8 -- wss://realtime.ably.io:443
2016-05-20 09:28:16.761 demo-ios[7894:67011] DEBUG: (ARTWebSocketTransport.m:231) R:0x7fd528da7d70 WS:0x7fd528c9e8b0 websocket did open
2016-05-20 09:28:16.761 demo-ios[7894:67011] DEBUG: (ARTWebSocketTransport.m:351) R:0x7fd528da7d70 WS:0x7fd528c9e8b0 websocket did receive data <85a66163 74696f6e 04ac636f 6e6e6563 74696f6e 4964aa6b 75717638 6c434c79 50ad636f 6e6e6563 74696f6e 4b6579d9 20393530 6138216b 75717638 6c434c79 50354246 74326b2d 62613133 39353061 38b0636f 6e6e6563 74696f6e 53657269 616cffb1 636f6e6e 65637469 6f6e4465 7461696c 7387a863 6c69656e 744964a1 2aad636f 6e6e6563 74696f6e 4b6579d9 20393530 6138216b 75717638 6c434c79 50354246 74326b2d 62613133 39353061 38ae6d61 784d6573 73616765 53697a65 cd1000ae 6d617849 6e626f75 6e645261 746505ac 6d617846 72616d65 53697a65 ce000400 00a87365 72766572 4964d927 66726f6e 74656e64 2e656466 352e6575 2d63656e 7472616c 2d312d41 2e692d31 61353963 386136b2 636f6e6e 65637469 6f6e5374 61746554 746cce00 01d4c0>
2016-05-20 09:28:16.762 demo-ios[7894:67011] VERBOSE: RS:0x7fd528d440c0 ARTJsonLikeEncoder<msgpack> decoding '<85a66163 74696f6e 04ac636f 6e6e6563 74696f6e 4964aa6b 75717638 6c434c79 50ad636f 6e6e6563 74696f6e 4b6579d9 20393530 6138216b 75717638 6c434c79 50354246 74326b2d 62613133 39353061 38b0636f 6e6e6563 74696f6e 53657269 616cffb1 636f6e6e 65637469 6f6e4465 7461696c 7387a863 6c69656e 744964a1 2aad636f 6e6e6563 74696f6e 4b6579d9 20393530 6138216b 75717638 6c434c79 50354246 74326b2d 62613133 39353061 38ae6d61 784d6573 73616765 53697a65 cd1000ae 6d617849 6e626f75 6e645261 746505ac 6d617846 72616d65 53697a65 ce000400 00a87365 72766572 4964d927 66726f6e 74656e64 2e656466 352e6575 2d63656e 7472616c 2d312d41 2e692d31 61353963 386136b2 636f6e6e 65637469 6f6e5374 61746554 746cce00 01d4c0>'; got: {
    action = 4;
    connectionDetails =     {
        clientId = "*";
        connectionKey = "950a8!kuqv8lCLyP5BFt2k-ba13950a8";
        connectionStateTtl = 120000;
        maxFrameSize = 262144;
        maxInboundRate = 5;
        maxMessageSize = 4096;
        serverId = "frontend.edf5.eu-central-1-A.i-1a59c8a6";
    };
    connectionId = kuqv8lCLyP;
    connectionKey = "950a8!kuqv8lCLyP5BFt2k-ba13950a8";
    connectionSerial = "-1";
}
2016-05-20 09:28:16.762 demo-ios[7894:67011] VERBOSE: RS:0x7fd528d440c0 ARTJsonLikeEncoder<msgpack>: protocolMessageFromDictionary {
    action = 4;
    connectionDetails =     {
        clientId = "*";
        connectionKey = "950a8!kuqv8lCLyP5BFt2k-ba13950a8";
        connectionStateTtl = 120000;
        maxFrameSize = 262144;
        maxInboundRate = 5;
        maxMessageSize = 4096;
        serverId = "frontend.edf5.eu-central-1-A.i-1a59c8a6";
    };
    connectionId = kuqv8lCLyP;
    connectionKey = "950a8!kuqv8lCLyP5BFt2k-ba13950a8";
    connectionSerial = "-1";
}
2016-05-20 09:28:16.763 demo-ios[7894:67011] VERBOSE: R:0x7fd528da7d70 ARTRealtime didReceive Protocol Message Connected 
2016-05-20 09:28:16.763 demo-ios[7894:67011] DEBUG: (ARTRealtime.m:212) R:0x7fd528da7d70 transition to Connected requested
2016-05-20 09:28:16.763 demo-ios[7894:67011] DEBUG: (ARTWebSocketTransport.m:68) R:0x7fd528da7d70 WS:0x7fd528c9e8b0 sending action 10 with (null)
2016-05-20 09:28:16.763 demo-ios[7894:67011] VERBOSE: RS:0x7fd528d440c0 ARTJsonLikeEncoder<msgpack>: protocolMessageToDictionary {
    action = 10;
    channel = "mobile:chat";
    msgSerial = 0;
}
2016-05-20 09:28:16.764 demo-ios[7894:67011] VERBOSE: RS:0x7fd528d440c0 ARTJsonLikeEncoder<msgpack> encoding '{
    action = 10;
    channel = "mobile:chat";
    msgSerial = 0;
}'; got: <83a76368 616e6e65 6cab6d6f 62696c65 3a636861 74a66163 74696f6e 0aa96d73 67536572 69616c00>
2016-05-20 09:28:16.806 demo-ios[7894:67011] DEBUG: (ARTWebSocketTransport.m:351) R:0x7fd528da7d70 WS:0x7fd528c9e8b0 websocket did receive data <83a66163 74696f6e 0ba76368 616e6e65 6cab6d6f 62696c65 3a636861 74ad6368 616e6e65 6c536572 69616cb0 61373332 39323831 32343936 363a2d31>
2016-05-20 09:28:16.806 demo-ios[7894:67011] VERBOSE: RS:0x7fd528d440c0 ARTJsonLikeEncoder<msgpack> decoding '<83a66163 74696f6e 0ba76368 616e6e65 6cab6d6f 62696c65 3a636861 74ad6368 616e6e65 6c536572 69616cb0 61373332 39323831 32343936 363a2d31>'; got: {
    action = 11;
    channel = "mobile:chat";
    channelSerial = "a732928124966:-1";
}
2016-05-20 09:28:16.807 demo-ios[7894:67011] VERBOSE: RS:0x7fd528d440c0 ARTJsonLikeEncoder<msgpack>: protocolMessageFromDictionary {
    action = 11;
    channel = "mobile:chat";
    channelSerial = "a732928124966:-1";
}
2016-05-20 09:28:16.807 demo-ios[7894:67011] VERBOSE: R:0x7fd528da7d70 ARTRealtime didReceive Protocol Message Attached 
2016-05-20 09:28:16.807 demo-ios[7894:67011] DEBUG: (ARTWebSocketTransport.m:68) R:0x7fd528da7d70 WS:0x7fd528c9e8b0 sending action 14 with (null)
2016-05-20 09:28:16.807 demo-ios[7894:67011] VERBOSE: RS:0x7fd528d440c0 ARTJsonLikeEncoder<msgpack>: presenceMessageToDictionary {
    action = 2;
}
2016-05-20 09:28:16.808 demo-ios[7894:67011] VERBOSE: RS:0x7fd528d440c0 ARTJsonLikeEncoder<msgpack>: protocolMessageToDictionary {
    action = 14;
    channel = "mobile:chat";
    msgSerial = 0;
    presence =     (
                {
            action = 2;
        }
    );
}
2016-05-20 09:28:16.808 demo-ios[7894:67011] VERBOSE: RS:0x7fd528d440c0 ARTJsonLikeEncoder<msgpack> encoding '{
    action = 14;
    channel = "mobile:chat";
    msgSerial = 0;
    presence =     (
                {
            action = 2;
        }
    );
}'; got: <84a96d73 67536572 69616c00 a8707265 73656e63 659181a6 61637469 6f6e02a7 6368616e 6e656cab 6d6f6269 6c653a63 686174a6 61637469 6f6e0e>
2016-05-20 09:28:16.842 demo-ios[7894:67011] DEBUG: (ARTWebSocketTransport.m:351) R:0x7fd528da7d70 WS:0x7fd528c9e8b0 websocket did receive data <84a66163 74696f6e 02a5636f 756e7401 a5657272 6f7283a7 6d657373 616765d9 454d616c 666f726d 6564206d 65737361 67653b20 6e6f2063 6c69656e 74496420 696e206d 65737361 6765206f 7220696e 66657272 65642066 726f6d20 636f6e6e 65637469 6f6ea463 6f6465cd 9c4caa73 74617475 73436f64 65cd0190 a96d7367 53657269 616c00>
2016-05-20 09:28:16.842 demo-ios[7894:67011] VERBOSE: RS:0x7fd528d440c0 ARTJsonLikeEncoder<msgpack> decoding '<84a66163 74696f6e 02a5636f 756e7401 a5657272 6f7283a7 6d657373 616765d9 454d616c 666f726d 6564206d 65737361 67653b20 6e6f2063 6c69656e 74496420 696e206d 65737361 6765206f 7220696e 66657272 65642066 726f6d20 636f6e6e 65637469 6f6ea463 6f6465cd 9c4caa73 74617475 73436f64 65cd0190 a96d7367 53657269 616c00>'; got: {
    action = 2;
    count = 1;
    error =     {
        code = 40012;
        message = "Malformed message; no clientId in message or inferred from connection";
        statusCode = 400;
    };
    msgSerial = 0;
}
2016-05-20 09:28:16.843 demo-ios[7894:67011] VERBOSE: RS:0x7fd528d440c0 ARTJsonLikeEncoder<msgpack>: protocolMessageFromDictionary {
    action = 2;
    count = 1;
    error =     {
        code = 40012;
        message = "Malformed message; no clientId in message or inferred from connection";
        statusCode = 400;
    };
    msgSerial = 0;
}
2016-05-20 09:28:16.843 demo-ios[7894:67011] VERBOSE: R:0x7fd528da7d70 ARTRealtime didReceive Protocol Message Nack 
2016-05-20 09:28:16.843 demo-ios[7894:67011] VERBOSE: R:0x7fd528da7d70 ARTRealtime Protocol Message with error ARTErrorInfo with code 400, message: Malformed message; no clientId in message or inferred from connection 
2016-05-20 09:28:16.843 demo-ios[7894:67011] VERBOSE: R:0x7fd528da7d70 ARTRealtime NACK: msgSerial=0, count=1
2016-05-20 09:28:16.843 demo-ios[7894:67011] VERBOSE: R:0x7fd528da7d70 ARTRealtime NACK (before processing): pendingMessageStartSerial=0, pendingMessages=1
2016-05-20 09:28:16.852 demo-ios[7894:67011] VERBOSE: R:0x7fd528da7d70 ARTRealtime NACK (after processing): pendingMessageStartSerial=1, pendingMessages=0
```

**Comment by tcard** [comment]
@stannedelchev Are you setting a client ID somewhere? I'm probably missing something but I can't find anywhere setting [this](https://github.com/ably/demo-mobile-ios/blob/df2a43d5600da947dba5d0dd225f975890de8bcd/demo-ios/ChatViewController.swift#L17) to a value.

**Comment by stannedelchev** [comment]
@tcard Yup, it's set [on segueing from the previous viewcontroller](https://github.com/ably/demo-mobile-ios/blob/master/demo-ios/EditNameViewController.swift#L34).

**Comment by tcard** [comment]
@mattheworiordan I'm not sure whether this is expected or not.

The spec says "If the client library is authenticated but unidentified (i.e. clientId is a wildcard '\*' or client is anonymous), the enter request results in an error immediately" ([RTP8f](http://docs.ably.io/client-lib-development-guide/features/#RTP8f)). I'm not sure "clientId is a wildcard '\*'" is true if there's a non-wildcard `ClientOptions.clientId` but the `TokenDetails`/`ConnectionDetails` had a wildcard `clientId`. [RSA7a](http://docs.ably.io/client-lib-development-guide/features/#RSA7a) seems to imply that this situation is not identified (and thus this is not a bug).

But then, [RSA7b4](http://docs.ably.io/client-lib-development-guide/features/#RSA7b4) says that "the client does not have an identity but is allowed to assume an identity when performing operations". I'm not sure if setting `ClientOptions.clientId` qualifies as a way of assuming an identity. It seems sensible to me but it isn't specified explicitly.

@stannedelchev In the meantime, `enterClient(clientId)` should work.

**Comment by paddybyers** [comment]
The connection has a wildcard `clientId`, as confirmed by the `CONNECTED` response, so all presence messages must then contain a specific `clientId` within the message. We don't allow the `clientId` associated with the connection to change once the connection is established.

> I'm not sure if setting ClientOptions.clientId qualifies as a way of assuming an identity. It seems sensible to me but it isn't specified explicitly.

So setting `ClientOptions.clientId` post-connection has to be seen as a way of setting a default identity for operations, instead of as a way of changing the state of the connection. I remember us discussing whether or not this is legitimate, and the objection at the time was that we would then be permitting operations having an explicit `clientId` (ie via `enterClient()`) that is different from `ClientOptions.clientId`.This has potential for confusion and inconsistency in the API, since `ClientOptions.clientId` now has different meanings in different contexts; sometimes it specifies the `clientId` to be used for the connection, and sometimes not, and sometimes you can perform operations for different `clientId`s.

I do agree that it seems reasonable to be able to use it as a default for operations, but I worry about the inconsistency.

**Comment by tcard** [comment]
@paddybyers 

> The connection has a wildcard clientId, as confirmed by the CONNECTED response, so all presence messages must then contain a specific clientId within the message

Right, my question is whether the library should set this specific clientId from the clientId passed at ClientOptions for presence messages.

> So setting ClientOptions.clientId post-connection has to be seen as a way of setting a default identity for operations

We don't allow this to happen, anyway. The ClientOptions passed to Realtime is copied and encapsulated and you can't change it from the outside (and the inside doesn't change it, of course).

> we would then be permitting operations having an explicit clientId (ie via enterClient()) that is different from ClientOptions.clientId

Don't we already allow that when publishing messages? In any case, if we want to disallow that, wouldn't it make more sense to disallow using `enterClient` if there is a `ClientOptions.clientId`? Setting `ClientOptions.clientId = foo` and then having to call `enterClient(foo)` anyway seems counterintuitive and unergonomic.

**Comment by mattheworiordan** [comment]
> Right, my question is whether the library should set this specific clientId from the clientId passed at ClientOptions for presence messages.

No, I don't see why that is ever necessary.  My understanding is that:

* `ClientOptions.clientId` can only be specified when the library is being instanced.  It should be immutable later, if the language permits, so changing it later should never have an effect.  If the `ConnectionDetails.clientId` returned is incompatible with the `clientId` configured in the library, then it should result in an error at that point and the connection should move to the `FAILED` state.
* If the `clientId` is configured as a wildcard, then entering without specifying a `clientId` is going to fail.
* If the `clientId` is configured as anything but a wildcard, then entering does not require a `clientId` to be sent as it is implicit.

>> we would then be permitting operations having an explicit clientId (ie via enterClient()) that is different from ClientOptions.clientId

> Don't we already allow that when publishing messages? In any case, if we want to disallow that, wouldn't it make more sense to disallow using enterClient if there is a ClientOptions.clientId? Setting ClientOptions.clientId = foo and then having to call enterClient(foo) anyway seems counterintuitive and unergonomic.

Yes, agreed, although in most libraries I believe `enterClient` shares the same code with `enter`, so as long as the `clientId` passed is compatible, it will work.  I don't think it's that important that the client library ensures compatibility though, although it would be nice.

**Comment by tcard** [comment]
> If the clientId is configured as a wildcard, then entering without specifying a clientId is going to fail.
> If the clientId is configured as anything but a wildcard, then entering does not require a clientId to be sent as it is implicit.

@mattheworiordan Well, the clientId _is_ specified, but at the `ClientOptions` instead, not in the argument to `enterClient`, and thus it can't be implicit in the connection. The situation here is:

* The dev sets `ClientOptions.clientId` to `foo`.
* The dev sets an auth URL that returns a `TokenDetails` with `clientId = "*"`
* The dev does `presence.enter`, which fails. The connection is _not_ configured for `clientId = "foo"` (since the token has a wildcard), but she expects `presence.enter` to implicitly (at the client library level) set the presence message's `clientId` to `foo`.

Is this expectation correct? Your comment wasn't so conclusive.

**Comment by mattheworiordan** [comment]
I would expect step 2 fails.

Looking at http://docs.ably.io/client-lib-development-guide/features/#RSA15b, I think it needs to be changed to:

```
If the clientId from TokenDetails or connectionDetails contains only a wildcard string '*', then the client is permitted to be either unidentified (i.e. anonymous without a clientId) or identified by providing an explicit clientId for operations that support a clientId argument such as publish or presence#enterClient.
```

Then I think we should change http://docs.ably.io/client-lib-development-guide/features/#RSA15c to:

```
Following an auth request which uses a TokenDetails or TokenRequest object that contains an incompatible clientId, the library should in the case of Realtime change the connection state to FAILED and emit an error, and in the case of REST result in an appropriate error response. A ClientOptions#clientId is considered incompatible when it has a string value, yet the clientId in the Token is any other value including a wildcard "*"
```

Oddly this relates to a separate issue https://github.com/ably/docs/issues/117#issuecomment-221221427.

@paddybyers WDYT.  I feel that providing a wildcard `clientId` in the token, yet explicitly providing one in the client library when instancing it is a recipe for disaster.  Saying that, I could understand why people may do that.  If we wanted to support this, then we should insist the `clientId` is in fact included in the WebSocket URL, and https://github.com/ably/docs/issues/117 should be closed.

**Comment by paddybyers** [comment]
> If we wanted to support this, then we should insist the clientId is in fact included in the WebSocket URL, and ably/docs#117 should be closed.

Yes, this was my point: `clientId` as a connection param is there so that the connection can be associated with a specific `clientId`, even if it cannot be inferred from the credentials.

**Comment by tcard** [comment]
Step 2 failing sounds good to me too.

@stannedelchev Thanks for spotting this! We're forbidding using wildcard tokens when `clientId` is set. So [this line](https://github.com/ably/demo-mobile-ios/blob/df2a43d5600da947dba5d0dd225f975890de8bcd/demo-ios/ChatModel.swift#L28) won't be correct. Use `presence.enterClient(self.clientId, data: nil)`in [this line](https://github.com/ably/demo-mobile-ios/blob/df2a43d5600da947dba5d0dd225f975890de8bcd/demo-ios/ChatModel.swift#L88) instead.
